test/common_nn.py

```diff
@@ -2501,6 +2601,45 @@ def ctcloss_reference(log_probs, targets, input_lengths, target_lengths, blank=0
     output = output.to(dt)
     return output
 
+
+def padding1d_circular(input, i):
+    r""" input:
+            [[[0., 1., 2.],
+              [3., 4., 5.]]]
+          i: 1
+          output:
+            [[[2., 0., 1., 2., 0.],
+              [5., 3., 4., 5., 3.]]]
+    """
+    return torch.cat([input[:, :, -i:], input,
+                      input[:, :, 0:i]], dim=2)
+
+
+def padding2d_circular(input, i):
+    r"""input:
+             [[[[0., 1.],
+                [2., 3.]]]]
+            i: 1
+    output:
+        [[[[3., 2., 3., 2.],
+           [1., 0., 1., 0.],
+           [3., 2., 3., 2.],
+           [1., 0., 1., 0.]]]]
+    """
+    input = torch.cat([input[:, :, :, -i:], input, input[:, :, :, 0:i]], dim=3)
+    return torch.cat([input[:, :, -i:], input, input[:, :, 0:i]], dim=2)
+
+
+def padding3d_circular(input, i):
+    r"""input:
+            [[[[[0, 1, 2]]]]]
+        i:  1
+    output: [[[[[2, 0, 1, 2, 0]]]]]
+    """
+    return torch.cat([input[:, :, :, :, -i:], input,
+                      input[:, :, :, :, 0:i]], dim=4)
```

Comment:
This looks wrong. But there is some fishy business about all of these reference implementations. See above.

Reply:
It didn't effect specific examples that I had but in general you're right. It has to cut dim=2 and dim=3. I will add it.

Reply:
addressed